Bundle policy checks in convenient function for widget checks

raethlein opened the pull request:
## Describe your changes

We run the following checks for all widgets:
- `check_fragment_path_policy`
- `check_cache_replay_rules`
- `check_callback_rules`
- `check_session_state_rules` 

This PR introduces that instead of calling them individually, we call a convenient function. This prevents missing any checks for new widgets.
Also, the `check_fragment_path` call was missing in `select_slider` and `download_button` and `dataframe`.

## GitHub Issue Link (if applicable)

## Testing Plan

- Unit Tests (JS and/or Python)
 - Add a unit test to ensure that all relevant path policies are called when calling the new function
 - Add missing check_fragment_path checks for `select_slider` and `download_button` and `dataframe`
---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

raethlein committed:
Add convenient function to check widget policies

raethlein committed:
Add convenient function for check all widget policies

lukasmasuch commented on lib/streamlit/elements/lib/policies.py:
```diff
@@ -30,7 +30,8 @@
 
 def check_callback_rules(dg: DeltaGenerator, on_change: WidgetCallback | None) -> None:
-    """Ensures that widgets other than `st.form_submit` within a form don't have an on_change callback set.
+    """Ensures that widgets other than `st.form_submit` within a form don't have an \
```
nit: `st.form_submit` doesn't exist, probably referring to `st.form_submit_button`?

lukasmasuch commented on lib/streamlit/elements/lib/policies.py:
```diff
@@ -51,9 +52,12 @@ def check_callback_rules(dg: DeltaGenerator, on_change: WidgetCallback | None) -
     default_value: Any, key: str | None, writes_allowed: bool = True
 ) -> None:
-    """Ensures that no values are set for widgets with the given key when writing is not allowed.
+    """Ensures that no values are set for widgets with the given key when writing \
```
Is the `\` at the ned of the docstring really needed here? I also remember that there was a problem with putting that into the string (related to the docstring formatting).

vdonato replied:
For docstrings, I think you can just wrap the line, and most tools that parse docstrings won't consider the docstring summary ended until they see a double return.

raethlein replied:
yep good point, its not needed 👍

lukasmasuch commented on lib/streamlit/elements/lib/policies.py:
```diff
@@ -71,7 +75,8 @@ def check_session_state_rules(
     if not writes_allowed:
         raise StreamlitAPIException(
-            f'Values for the widget with key "{key}" cannot be set using `st.session_state`.'
+            f'Values for the widget with key "{key}" cannot be set using \
```
I think we usually just do this:

```
raise StreamlitAPIException(
   f'Values for the widget with key "{key}" cannot be set using'
   ' `st.session_state`.'
)
```

But I'm not really aware of the implications / down/upsides of using this way for multiline string

vdonato replied:
+1 to using implicit string concatenation where possible

lukasmasuch reviewed (commented):
LGTM 👍 Thanks for cleaning this up! But two aspects to look into:

1. I'm a bit worried about the usage of the `/` for multiline strings. Mainly because we haven't yet used that as far as I'm aware of and it already caused some trouble with the ruff docstring formatter. Not sure how well this works together with our docstring formatter or if it introduces other unexpected side-effects. And I think there isn't really a need to use this.
2. The elements with selections also use these checks and I believe can be updated to use this function as well:

https://github.com/streamlit/streamlit/blob/3aad5e2a97998f56d4fd936f6d6269960426720d/lib/streamlit/elements/arrow.py#L492-L495

vdonato reviewed (approved):
+1 to Lukas' comments, but otherwise this LGTM

raethlein committed:
Use implict string concatenation

raethlein committed:
Use check_widget_policies for selection widgets

raethlein commented:
> The elements with selections also use these checks and I believe can be updated to use this function as well:

great point! I have added it. Also, the `fragment_path` check was missing for `dataframe`, which is now added 👍
